File: elixir/lib/bonemesh/cert.ex
defmodule Bonemesh.Cert do
  @moduledoc """
  A BoneMesh v3 membership certificate (security.md §3): a mesh-root-signed
  binding of a display label to a node's ML-DSA-65 identity key, valid within a
  time window. Represented as a plain map with string keys (the JSON form); its
  signed pre-image is the `Bonemesh.Canon` canonicalization of every field
  except `"sig"`. Interoperates with the Java reference (shared corpus and the
  PQC interop vector).
  """

  @version 3

  @doc """
  Builds an unsigned certificate map.

  `identity_key` is the node's raw ML-DSA-65 public key.
  """
  @spec new(String.t(), String.t(), binary(), integer(), integer()) :: map()
  def new(mesh, label, identity_key, not_before, not_after) do
    %{
      "v" => @version,
      "mesh" => mesh,
      "label" => label,
      "idk" => Base.encode64(identity_key),
      "nbf" => not_before,
      "exp" => not_after
    }
  end

  @doc "Signs a certificate with the mesh root's ML-DSA-87 private key."
  @spec sign(map(), binary()) :: map()
  def sign(cert, root_private) do
    sig = Bonemesh.Crypto.mldsa_sign(:mldsa87, root_private, Bonemesh.Canon.canonicalize(cert))
    Map.put(cert, "sig", Base.encode64(sig))
  end

  @doc """
  Verifies a certificate against the pinned root public key, mesh, and time.
  Returns `:ok` or `{:error, reason}`.
  """
  @spec verify(map(), binary(), String.t(), integer()) :: :ok | {:error, String.t()}
  def verify(cert, root_public, expected_mesh, now) do
    cond do
      cert["mesh"] != expected_mesh -> {:error, "mesh mismatch"}
      now < cert["nbf"] -> {:error, "certificate not yet valid"}
      now > cert["exp"] -> {:error, "certificate expired"}
      not Map.has_key?(cert, "sig") -> {:error, "certificate is unsigned"}
      true -> verify_signature(cert, root_public)
    end
  end

  @doc "The node's raw ML-DSA-65 identity public key."
  @spec identity_key(map()) :: binary()
  def identity_key(cert), do: Base.decode64!(cert["idk"])

  defp verify_signature(cert, root_public) do
    sig = Base.decode64!(cert["sig"])
    pre_image = cert |> Map.delete("sig") |> Bonemesh.Canon.canonicalize()

    if Bonemesh.Crypto.mldsa_verify(:mldsa87, root_public, pre_image, sig),
      do: :ok,
      else: {:error, "root signature does not verify"}
  end
end

File: python/bonemesh/cert.py
"""BoneMesh v3 membership certificate (security.md §3).

A mesh-root-signed binding of a label to a node's ML-DSA-65 identity key. The
JSON form is a plain dict; its signed pre-image is the canonicalization of every
field except ``sig``. Base64 fields use the standard alphabet with padding.
"""

from __future__ import annotations

import base64
import binascii

from bonemesh.canon import canonicalize
from bonemesh.crypto import mldsa87_sign, mldsa87_verify


def build(mesh: str, label: str, identity_key: bytes, not_before: int, not_after: int) -> dict:
    """An unsigned certificate. ``identity_key`` is the raw ML-DSA-65 public key."""
    return {
        "v": 3,
        "mesh": mesh,
        "label": label,
        "idk": base64.b64encode(identity_key).decode("ascii"),
        "nbf": not_before,
        "exp": not_after,
    }


def sign(cert: dict, root_private_seed: bytes) -> dict:
    """Returns a copy of ``cert`` carrying the mesh root's ML-DSA-87 signature.

    Used by the test suite and by tooling that issues certificates; a node itself
    only ever verifies.
    """
    signed = {k: v for k, v in cert.items() if k != "sig"}
    signed["sig"] = base64.b64encode(
        mldsa87_sign(root_private_seed, canonicalize(signed))
    ).decode("ascii")
    return signed


def verify(cert: dict, root_public: bytes, expected_mesh: str, now: int) -> str | None:
    """Returns None if valid, else a reason string."""
    if cert.get("mesh") != expected_mesh:
        return "mesh mismatch"
    nbf = cert.get("nbf")
    # bool is an int in Python; a certificate carrying `true` is malformed, not
    # a timestamp of 1.
    if not isinstance(nbf, int) or isinstance(nbf, bool) or now < nbf:
        return "certificate not yet valid"
    exp = cert.get("exp")
    if not isinstance(exp, int) or isinstance(exp, bool) or now > exp:
        return "certificate expired"
    if not isinstance(cert.get("sig"), str):
        return "certificate is unsigned"
    try:
        sig = base64.b64decode(cert["sig"], validate=True)
    except (binascii.Error, ValueError):
        return "signature is not base64"
    try:
        pre_image = canonicalize(cert)
    except ValueError as e:
        return f"canon: {e}"
    if not mldsa87_verify(root_public, pre_image, sig):
        return "root signature does not verify"
    return None


def identity_key(cert: dict) -> bytes:
    """The node's raw ML-DSA-65 public key."""
    return base64.b64decode(cert["idk"])
